File: tkinterapp1/game.py
# -*- coding: utf-8 -*-
"""推箱子核心（Tkinter / PyQt 等 Python GUI 共用教学逻辑）。"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Set, Tuple


def cell_key(x: int, y: int) -> str:
    return f"{x},{y}"


@dataclass
class Hist:
    player: Tuple[int, int]
    box_from: Optional[str] = None
    box_to: Optional[str] = None


@dataclass
class GameState:
    walls: Set[str] = field(default_factory=set)
    goals: Set[str] = field(default_factory=set)
    boxes: Set[str] = field(default_factory=set)
    player: Tuple[int, int] = (0, 0)
    moves: int = 0
    won: bool = False
    width: int = 0
    height: int = 0
    hist: List[Hist] = field(default_factory=list)

    @classmethod
    def from_rows(cls, rows: List[str]) -> "GameState":
        s = cls()
        max_x = max_y = 0
        for y, row in enumerate(rows):
            max_y = y
            for x, ch in enumerate(row):
                max_x = max(max_x, x)
                k = cell_key(x, y)
                if ch == "#":
                    s.walls.add(k)
                elif ch == ".":
                    s.goals.add(k)
                elif ch == "$":
                    s.boxes.add(k)
                elif ch == "*":
                    s.boxes.add(k)
                    s.goals.add(k)
                elif ch == "@":
                    s.player = (x, y)
                elif ch == "+":
                    s.player = (x, y)
                    s.goals.add(k)
        s.width = max_x + 1
        s.height = max_y + 1
        return s

    def try_move(self, dx: int, dy: int) -> bool:
        if self.won:
            return False
        px, py = self.player
        nx, ny = px + dx, py + dy
        nk = cell_key(nx, ny)
        if nk in self.walls:
            return False
        if nk in self.boxes:
            bx, by = nx + dx, ny + dy
            bk = cell_key(bx, by)
            if bk in self.walls or bk in self.boxes:
                return False
            self.hist.append(Hist(self.player, nk, bk))
            self.boxes.discard(nk)
            self.boxes.add(bk)
            self.player = (nx, ny)
            self.moves += 1
            self.won = all(b in self.goals for b in self.boxes)
            return True
        self.hist.append(Hist(self.player))
        self.player = (nx, ny)
        return True

    def undo(self) -> bool:
        if self.won or not self.hist:
            return False
        entry: Optional[Hist] = None
        while self.hist:
            entry = self.hist.pop()
            if entry.box_from is not None:
                break
            self.player = entry.player
        if entry is None or entry.box_from is None:
            return True
        self.player = entry.player
        assert entry.box_to is not None
        self.boxes.discard(entry.box_to)
        self.boxes.add(entry.box_from)
        if self.moves > 0:
            self.moves -= 1
        self.won = False
        return True


LEVEL = [
    "#######",
    "#. . .#",
    "# $$$ #",
    "#.$@$.#",
    "# $$$ #",
    "#. . .#",
    "#######",
]

File: mauiapp1/Game.cs
// 推箱子核心（.NET MAUI 教学）
namespace SokobanMaui;

public sealed class GameState
{
    public HashSet<string> Walls { get; } = new();
    public HashSet<string> Goals { get; } = new();
    public HashSet<string> Boxes { get; } = new();
    public (int X, int Y) Player { get; private set; }
    public int Moves { get; private set; }
    public bool Won { get; private set; }
    public int Width { get; private set; }
    public int Height { get; private set; }
    private readonly List<Hist> _hist = new();
    private readonly record struct Hist((int X, int Y) Player, string? BoxFrom, string? BoxTo);

    public static string Key(int x, int y) => $"{x},{y}";

    public static GameState FromRows(string[] rows)
    {
        var s = new GameState();
        int maxX = 0, maxY = 0;
        for (int y = 0; y < rows.Length; y++)
        {
            maxY = y;
            for (int x = 0; x < rows[y].Length; x++)
            {
                if (x > maxX) maxX = x;
                var k = Key(x, y);
                switch (rows[y][x])
                {
                    case '#': s.Walls.Add(k); break;
                    case '.': s.Goals.Add(k); break;
                    case '$': s.Boxes.Add(k); break;
                    case '*': s.Boxes.Add(k); s.Goals.Add(k); break;
                    case '@': s.Player = (x, y); break;
                    case '+': s.Player = (x, y); s.Goals.Add(k); break;
                }
            }
        }
        s.Width = maxX + 1;
        s.Height = maxY + 1;
        return s;
    }

    public bool TryMove(int dx, int dy)
    {
        if (Won) return false;
        var (px, py) = Player;
        int nx = px + dx, ny = py + dy;
        var nk = Key(nx, ny);
        if (Walls.Contains(nk)) return false;
        if (Boxes.Contains(nk))
        {
            int bx = nx + dx, by = ny + dy;
            var bk = Key(bx, by);
            if (Walls.Contains(bk) || Boxes.Contains(bk)) return false;
            _hist.Add(new Hist(Player, nk, bk));
            Boxes.Remove(nk);
            Boxes.Add(bk);
            Player = (nx, ny);
            Moves++;
            Won = Boxes.All(b => Goals.Contains(b));
            return true;
        }
        _hist.Add(new Hist(Player, null, null));
        Player = (nx, ny);
        return true;
    }

    public bool Undo()
    {
        if (Won || _hist.Count == 0) return false;
        Hist? e = null;
        while (_hist.Count > 0)
        {
            e = _hist[^1];
            _hist.RemoveAt(_hist.Count - 1);
            if (e.Value.BoxFrom != null) break;
            Player = e.Value.Player;
        }
        if (e is null || e.Value.BoxFrom is null) return true;
        Player = e.Value.Player;
        Boxes.Remove(e.Value.BoxTo!);
        Boxes.Add(e.Value.BoxFrom);
        if (Moves > 0) Moves--;
        Won = false;
        return true;
    }

    public char CellAt(int x, int y)
    {
        var k = Key(x, y);
        if (Player == (x, y)) return Goals.Contains(k) ? '+' : '@';
        if (Boxes.Contains(k)) return Goals.Contains(k) ? '*' : '$';
        if (Walls.Contains(k)) return '#';
        if (Goals.Contains(k)) return '.';
        return ' ';
    }

    public string RenderBoard()
    {
        var lines = new List<string>();
        for (int y = 0; y < Height; y++)
        {
            var row = new char[Width];
            for (int x = 0; x < Width; x++)
                row[x] = CellAt(x, y) == ' ' ? '·' : CellAt(x, y);
            lines.Add(new string(row));
        }
        return string.Join("\n", lines);
    }
}

public static class Levels
{
    public static readonly string[] Mini =
    {
        "#######", "#. . .#", "# $$$ #", "#.$@$.#",
        "# $$$ #", "#. . .#", "#######",
    };
}
